### PR TITLE
created analysis module to bundle up various methods

### PR DESCRIPTION
**Pull request by nikokaoja**
Moved set of methods that were in _analysis module to be class methods.
Renamed methods (usually dropped "to" and "get")

This is prep for adding methods to the analysis class that will analyze rules to detect modeling scenarion.

### CODE REVIEW

File: cognite/neat/rules/_analysis/_base.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from abc import ABC, abstractmethod
+
+from cognite.neat.rules._shared import Rules
+from cognite.neat.rules.models._rules._types import ClassEntity
+
+
+class BaseAnalysis(ABC):
+    @abstractmethod
+    def subset_rules(self, desired_classes: set[ClassEntity]) -> Rules:
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_base.py:9]: A good use case for generics. 
```suggestion
from typing import TypeVar

T_Rules = TypeVar("T_Rules", bound=Rules)

class BaseAnalysis(ABC, Generic[T_Rules]):
    @abstractmethod
    def subset_rules(self, desired_classes: set[ClassEntity]) -> T_Rules:
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        for _, row in class_linkage.iterrows():
+            source = row.source_class
+            target = row.target_class
+            target_targets = class_linkage[class_linkage.source_class == target].target_class.values
+            if source in target_targets and (source, target) not in sym_pairs:
+                sym_pairs.add((source, target))
+        return sym_pairs
+
+    def as_property_dict(self) -> dict[str, list[InformationProperty]]:
+        """This is used to capture all definitions of a property in the data model."""
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:251]: Is it possible to have multiple definition of the same property? Shouldn't that be an error?

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        """
+        return self.defined_classes(consider_inheritance) - self.connected_classes(consider_inheritance)
+
+    def symmetrically_connected_classes(
+        self, consider_inheritance: bool = False
+    ) -> set[tuple[ClassEntity, ClassEntity]]:
+        """Returns a set of pairs of symmetrically linked classes.
+
+        Args:
+            rules: Instance of InformationRules holding the data model
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:228]:
```suggestion
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+        Returns:
+            Set of classes that are disconnected from other classes
+        """
+        return self.defined_classes(consider_inheritance) - self.connected_classes(consider_inheritance)
+
+    def symmetrically_connected_classes(
+        self, consider_inheritance: bool = False
+    ) -> set[tuple[ClassEntity, ClassEntity]]:
+        """Returns a set of pairs of symmetrically linked classes.
+
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:226]: A sentence of what a symmetrically connected class is would be helpful.

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        class_property_pairs = self.classes_with_properties(consider_inheritance)
+        properties = list(itertools.chain.from_iterable(class_property_pairs.values()))
+
+        return {property.class_ for property in properties}
+
+    def disconnected_classes(self, consider_inheritance: bool = False) -> set[ClassEntity]:
+        """Return a set of classes that are disconnected (i.e. isolated) from other classes.
+
+        Args:
+            rules: Instance of InformationRules holding the data model
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:214]:
```suggestion
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+            Set of classes that are connected to other classes
+        """
+        class_linkage = self.class_linkage(consider_inheritance)
+        return set(class_linkage.source_class.values).union(set(class_linkage.target_class.values))
+
+    def defined_classes(self, consider_inheritance: bool = False) -> set[ClassEntity]:
+        """Returns classes that have properties defined for them in the data model.
+
+        Args:
+            rules: Instance of InformationRules holding the data model
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:199]:
```suggestion
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        class_linkage.drop_duplicates(inplace=True)
+        class_linkage = class_linkage[["source_class", "connecting_property", "target_class", "max_occurrence"]]
+
+        return class_linkage
+
+    def connected_classes(self, consider_inheritance: bool = False) -> set[ClassEntity]:
+        """Return a set of classes that are connected to other classes.
+
+        Args:
+            rules: Instance of InformationRules holding the data model
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:186]:
```suggestion
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+                    }
+                )
+                class_linkage = pd.concat([class_linkage, new_row.to_frame().T], ignore_index=True)
+
+        class_linkage.drop_duplicates(inplace=True)
+        class_linkage = class_linkage[["source_class", "connecting_property", "target_class", "max_occurrence"]]
+
+        return class_linkage
+
+    def connected_classes(self, consider_inheritance: bool = False) -> set[ClassEntity]:
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:182]: Can't there be several subset of connected classes? I would expect this to return
```suggestion
    def connected_classes(self, consider_inheritance: bool = False) -> list[set[ClassEntity]]:
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+                    processed_properties[property_.property_] = property_
+            class_property_pairs[class_] = processed_properties
+
+        return class_property_pairs
+
+    def class_linkage(self, consider_inheritance: bool = False) -> pd.DataFrame:
+        """Returns a dataframe with the class linkage of the data model.
+
+        Args:
+            rules: Instance of InformationRules holding the data model
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:153]:
```suggestion
```

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+                    )
+                    continue
+
+                if (only_rdfpath and property_.rule_type == TransformationRuleType.rdfpath) or not only_rdfpath:
+                    processed_properties[property_.property_] = property_
+            class_property_pairs[class_] = processed_properties
+
+        return class_property_pairs
+
+    def class_linkage(self, consider_inheritance: bool = False) -> pd.DataFrame:
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:149]: I know this is copy-pasted. But I think we should add some tech debt to remove the `pd.DataFrame` here, unless there is a specific reason for it.

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+
+        if consider_inheritance:
+            class_parent_pairs = self.class_parent_pairs()
+            for class_ in class_parent_pairs:
+                self._add_inherited_properties(class_, class_property_pairs, class_parent_pairs)
+
+        return class_property_pairs
+
+    @classmethod
+    def _add_inherited_properties(cls, class_, class_property_pairs, class_parent_pairs):
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:79]: type hints.

File: cognite/neat/rules/_analysis/_information_rules.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+            it will not be included in the returned dictionary.
+        """
+
+        class_property_pairs: dict[ClassEntity, list[InformationProperty]] = {}
+
+        for property_ in self.rules.properties:
+            if property_.class_ in class_property_pairs:
+                class_property_pairs[property_.class_] += [property_]
+            else:
+                class_property_pairs[property_.class_] = [property_]
```
Comment (doctrino) [inline cognite/neat/rules/_analysis/_information_rules.py:69]:
```suggestion
        class_property_pairs: dict[ClassEntity, list[InformationProperty]] = defaultdict(list)

        for property_ in self.rules.properties:
            class_property_pairs[property_.class_].append(property_)
```

File: tests/tests_unit/rules/test_analysis.py [hunk 13]
```diff
@@ -1,32 +1,40 @@
...
-    get_disconnected_classes,
-    get_symmetric_pairs,
-)
+from cognite.neat.rules._analysis._information_rules import InformationArchitectRulesAnalysis
 from cognite.neat.rules.models._rules import InformationRules
 from cognite.neat.rules.models._rules._types import ClassEntity
 
 
 class TestRulesAnalysis:
     def test_defined_classes(self, david_rules: InformationRules) -> None:
```
Comment (doctrino) [inline tests/tests_unit/rules/test_analysis.py:7]: Should probably add a few more tests here of the public methods in the `InformationArchitectRulesAnalysis`.

**Review by doctrino — APPROVED** [review]
Good job, like the structure.

File: cognite/neat/rules/_analysis/_information_rules.py (hunk 4 above)
Comment (nikokaoja) [reply cognite/neat/rules/_analysis/_information_rules.py:226]: Good catch ! Added

File: cognite/neat/rules/_analysis/_information_rules.py (hunk 9 above)
Comment (nikokaoja) [reply cognite/neat/rules/_analysis/_information_rules.py:153]: Removed all these args :)

File: tests/tests_unit/rules/test_analysis.py (hunk 13 above)
Comment (nikokaoja) [reply tests/tests_unit/rules/test_analysis.py:16]: done!

**Pull request merged**